sno/sno_repo.py

```diff
@@ -0,0 +1,146 @@
+    - there is property in the repo config called sno.repository.version that contains the dataset format version
+      number, which is used until the sno.repository.version blob is written.
+    - the .sno/index file has been extended to stop git messing things up - see LOCKED_EMPTY_GIT_INDEX.
+
+    There can also be properties in the repo config relating to the working copy - named sno.workingcopy.* - but
+    these are not written by this function, it is up to the caller to configure these as needed.
+    See WorkingCopy.write_config()
+    """
+
+    repo = pygit2.init_repository(str(repo_root_path.resolve()))
+    repo = convert_git_repo_to_sno_repo(repo_root_path)
+    write_repo_version_config(repo, repo_version)
+    assert is_sno_repo(repo)
+    return repo
+
+
+def convert_git_repo_to_sno_repo(repo_root_path):
+    """Given a newly created non-bare git repo, with the following structure:
+    repo_root_path/  # also worktree path
+        .git/
+            HEAD
+            ...      # objects, refs, etc
+    Converts it to a sno repo with the following structure:
+    repo_root_path/  # also worktree path, as before
+        .git         # contains "gitdir: .sno"
+        .sno/
+            HEAD
+            index    # Locked using .sno extension
+            ...      # etc
+    """
+
+    # Note: pygit2 can read the .sno directory okay, but it won't create it, so we do it ourselves using Path.rename()
+    # Note: We could also call `git init --separate-git-dir=.sno` to crete it but it doesn't work very well - it
+    # converts all paths to be absolute, which would mean sno repos wouldn't be renameable.
+
+    # Move .git to .sno but add a reference at .git so git can find it.
+    dot_git_path = repo_root_path / ".git"
+    assert dot_git_path.is_dir()
+    dot_sno_path = repo_root_path / ".sno"
+    assert not dot_sno_path.exists()
+
+    dot_git_path.rename(dot_sno_path)
+    dot_git_path.write_text("gitdir: .sno\n", encoding="utf-8")
+
+    repo = pygit2.Repository(str(repo_root_path.resolve()))
+    assert is_sno_repo(repo)
+
+    # Disable git commands that would mess things up:
+    repo_files.write_repo_file(repo, repo_files.INDEX, LOCKED_EMPTY_GIT_INDEX)
+
+    # Force writing to reflogs.
+    repo.config["core.logAllRefUpdates"] = "always"
+    return repo
+
+
+def is_sno_repo(pygit_repo):
+    if pygit_repo is not None:
+        # Older sno repos were just bare repos, newer sno repos are non-bare git repos that use ".sno" instead of ".git"
+        return pygit_repo.is_bare or Path(pygit_repo.path).stem == ".sno"
+    return False
+
+
+def get_sno_workdir(pygit_repo):
+    # Older sno repos are bare and use repo.path as the "workdir" - that is, where working copies are written -
+    # Newer ones have an actual workdir.
+    return Path(pygit_repo.path) if pygit_repo.is_bare else Path(pygit_repo.workdir)
+
+
+@contextlib.contextmanager
+def unlocked_git_index(pygit_repo):
```

Comment:
we should probably start to resurrect `fsck` again to allow us to do things like "re-lock" an index if this process crashes hard and the index doesn't get reset.

Reply:
Would prefer some more specific tests around locking and unlocking behaviours

Reply:
Added some more specific tests in test_sno_repo.py

If it seems best, we could eagerly re-lock the index every time we open up a SnoRepo, even if we expect it to be already locked - as opposed to now, where we leave it how it is until next time we need to unlock / relock it.

Reply:
Not sure about every time, feels like that'd mess up concurrent calls to `sno`?

Whatever we can come up with that is reasonably sane + safe :)